keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlArtifactRepository.kt

```diff
@@ -221,31 +231,53 @@ class SqlArtifactRepository(
     }
   }
 
-  override fun versions(artifact: DeliveryArtifact, limit: Int): List<String> =
-    if (isRegistered(artifact.name, artifact.type)) {
-      sqlRetry.withRetry(READ) {
-        jooq
-          .select(ARTIFACT_VERSIONS.VERSION, ARTIFACT_VERSIONS.RELEASE_STATUS)
-          .from(ARTIFACT_VERSIONS)
-          .where(ARTIFACT_VERSIONS.NAME.eq(artifact.name))
-          .and(ARTIFACT_VERSIONS.TYPE.eq(artifact.type))
-          .apply { if (artifact.statuses.isNotEmpty()) and(ARTIFACT_VERSIONS.RELEASE_STATUS.`in`(*artifact.statuses.map { it.toString() }.toTypedArray())) }
-          .fetch()
-          .getValues(ARTIFACT_VERSIONS.VERSION)
-      }
-        .sortedWith(artifact.versioningStrategy.comparator)
-        .also { versions ->
-          // FIXME: remove special handling for Docker
-          // FIXME: limit the sql query, not the returned list (once we'll have a native sorting mechanism)
-          return if (artifact is DockerArtifact) {
-            filterDockerVersions(artifact, versions, limit)
-          } else {
-            versions.subList(0, Math.min(versions.size, limit))
+  override fun versions(artifact: DeliveryArtifact, limit: Int): List<String> {
+    // sanity checks
+    if (!isRegistered(artifact.name, artifact.type)) {
+      throw NoSuchArtifactException(artifact)
+    }
+
+    if (artifact.sortBy == BRANCH_AND_TIMESTAMP && artifact.branch == null) {
+      throw InvalidArtifactSpecException(
+        "The ${artifact.type} artifact ${artifact.name} specifies sorting by branch and timestamp, but no branch is defined in the config.")
```

Comment:
```suggestion
        "Artifact ${artifact.reference} missing required branch for branch and timestamp sorting: ($artifact)")
```
I think using reference is better, and I like adding the config in the message so that it's clear what the current config is (to help with parsing issues/etc)

Reply:
Yeah, I wanted to print the delivery config name but wasn't 100% positive it would be (always) filled.